### ISSUE TITLE
Feature request: Add support for creating virtual hardware and virtual sensors/devices to dzVents

### ISSUE THREAD

**Issue by elxsi** [domoticz/domoticz#6211]
Hi :)

I'm currently developing a dzVents LuA script which queries some external devices for different sensor values (about 50 different sensors like temperatures, electricity, current and many others). It would be really nice to have some native Lua functions to directly create hardware, virtual sensors, and devices, rather than relying on the JSON API through domoticz.openURL(). Native Lua functions would provide a more seamless and secure integration within the Domoticz environment, reducing the complexity of managing external HTTP calls and ensuring compatibility across different setups. Additionally, this approach would eliminate potential issues with network connectivity or misconfigurations in API endpoints, while also offering better error handling and feedback mechanisms directly within the scripting context. Moreover, a native implementation would align with the principle of keeping Domoticz scripts self-contained and less dependent on external tools or processes, making them more robust and maintainable over time

So maybe it would be possible to extend the dz object with some new methods:

**Creation**
local hw = dz.createVirtualHardware(name, type, [port], [enabled]);

local dev = dz.createVirtualSensor(name, type, hwidx (or hwname));

local dev = dz.createVirtualDevice(name, deviceType, deviceSubType, hwidx (or hw name));

**Removal**
local result = dz.deleteVirtualHardware(hwname (or hwidx))

local result = dz.deleteVirtualSensor(name (or idx));

local result = dz.deleteVirtualDevice(name (or idx));

**Enable/Disable**
local result = dz.enableVirtualHardware(hwname (or hwidx)) -- and disableVirtualHardware

local result = dz.enableVirtualSensor(hwname (or hwidx)) -- and disableVirtualSensor

local result = dz.enableVirtualDevice(hwname (or hwidx)) -- and disableVirtualDevice

Yes, it would be possible to just create a LuA utility file defining the corresponding functions, but I think it would be a lot of better to directly include this functionality in the dzVents.

Thank you!

**Comment by elxsi** [comment]
In the meantime, I've added a LuA module to get some functions to create/delete hardware and devices (see attached files). The LuA module just uses the JSON API via curl or wget (openURL will be added later on).

The drawback is, that if hardware or devices are added via the JSON API, the domoticz dz object is not updated in the currently running LuA context - this is also the reason for this feature request and a native implementation. To overcome this problems in the LuA script, events can will be emitted (see dzExt_test.lua).

The dzExt.lua file must be placed in <domoticz>/scripts/lua/ and the test script must be placed in <domoticz>/scripts/dzVents/scripts/ (both files must be renamed to .lua, because .lua is not a valid extension for file uploads on Github):

[dzExt_test.txt](https://github.com/user-attachments/files/17792175/dzExt_test.txt)
[dzExt.txt](https://github.com/user-attachments/files/17792176/dzExt.txt)


Code can be used without any restrictions.

**Comment by waltervl** [comment]
I do not think this should be part of standard DzVents as you can have for example multiple Virtual device hardware gateways active. Also there should be no need to create devices from Dzvents as that should be arranged by the hardware gateway.

You have the possibility to put your helper functions in your shared helper library in global_data.lua on your system. Either in /path/to/domoticz/scripts/dzVents/script or in Domoticz’ internal GUI web editor. See for more info https://wiki.domoticz.com/DzVents:_next_generation_Lua_scripting#Shared_helper_functions

**Comment by elxsi** [comment]
Hi :)

> I do not think this should be part of standard DzVents as you can have for example multiple Virtual device hardware gateways active. 

I do not understand, why this might be an issue? You just call the createDevice function and provide the corresponding hardware name, so multiple virtual gateways are not a problem in my opinion.

> Also there should be no need to create devices from Dzvents as that should be arranged by the hardware gateway.

This might be true for hardware gateways, but not for virtual gateways. For example, I'm developing a LuA script which reads out data via ModBus RTU from a specific device. This device provides multiple sensors and values and I do not want to create all sensors manually, which is error prone.

> You have the possibility to put your helper functions in your shared helper library in global_data.lua on your system. Either in /path/to/domoticz/scripts/dzVents/script or in Domoticz’ internal GUI web editor. See for more info https://wiki.domoticz.com/DzVents:_next_generation_Lua_scripting#Shared_helper_functions

Yes, I know. But in this case, I've preferred to use /scripts/lua/ because I'd like to have a better separation and I do not want this to be automatically loaded in all of my LuA scripts.

It would be really nice to have native support for (at least) device creation in dzVents.

**Comment by waltervl** [comment]
I do not think Lua (and its derived language DzVents) scripting is designed to maintain hardware gateways. You better invest in the Python Plugin framewerk. That is designed to create and maintain custom hardware gateways. You also can invest in adding your Modbus RTU device as a C++ hardware gateway. 

There are by the way multiple Hardware gateways in Python and perhaps also in C++ See for example https://forum.domoticz.com/viewtopic.php?t=21297

Or search for modbus in https://wiki.domoticz.com/Plugins

**Comment by elxsi** [comment]
I understand, that maybe the creation of hardware is not the correct place in dzVents, but I still think, that it should be possible to create and manage devices/sensor with dzVents. The Modbus RTU part is already finished and working and is not part of this discussion.

**Comment by waltervl** [comment]
As we already have Python and C++ integration frameworks there should be no need for a third framework (which needs maintenance too). 
And still, If you want to add it to Dzvents it gives that possibility by using the dzvents helpers functionality as described earlier.

**Comment by elxsi** [comment]
Hey waltervl,

thank you for your response!

Currently, I am using the JSON API to create hardware and devices. While this generally works, it has several disadvantages, such as:

- Difficult error handling and generally prone to errors
- Poor performance (no native code execution)
- Cumbersome to use
- Delayed availability of created hardware/devices, as the dz object is not dynamically regenerated and updates only during the next script run

The existing Python plugin system is certainly nice but also has its drawbacks, such as:

- Higher development effort
- Occasional crashes
- Compatibility issues with different Python versions

For these reasons, I believe an integration for managing hardware/devices (also due to the simplicity of Lua) would be an excellent feature.

**Comment by akamming** [comment]
This should indeed be a great addition. I have several scripts for which i manually needed to create virtual sensors.   It would be create if the sensors could be created as part of the definition of the script.

So for now you have something like : 
```
return
{
    active = true, -- optional
    on =
    {
        -- at least one of these:
        customEvents = { ... },
        devices = { ... },
        groups = { ... },
        httpResponses = { ... },
        scenes = { ... },
        security = { ... },
        shellCommandResponses = { ... },
        system = { ... },
        timer = { ... },
        variables = { ... },

    },
    data = { ... }, -- optional
    logging = { ... }, -- optional
    execute = function(domoticz, item, triggerInfo)
    -- your code here
    end
}
``` 

But would be great if a Devices sections were added where you could specifiy your devices which are automatically created (and also automatically deleted if the script is deleted or the definition is removed), so something like: 
```
return
{
    active = true, -- optional
    VirtualDevices = { ... } 
    on =
    {
        -- at least one of these:
        customEvents = { ... },
        devices = { ... },
        groups = { ... },
        httpResponses = { ... },
        scenes = { ... },
        security = { ... },
        shellCommandResponses = { ... },
        system = { ... },
        timer = { ... },
        variables = { ... },

    },
    data = { ... }, -- optional
    logging = { ... }, -- optional
    execute = function(domoticz, item, triggerInfo)
    -- your code here
    end
}
```
Where in the customdevices sections, you could specify the kind of virtual devices you would be using in the script.

**Comment by waltervl** [comment]
@akamming as a python plugin author, why do you not use a python plugin for creating and maintaining devices?

**Comment by elxsi** [comment]
> This should indeed be a great addition. I have several scripts for which i manually needed to create virtual sensors. It would be create if the sensors could be created as part of the definition of the script.
> 
> So for now you have something like :
> 
> ```
> return
> {
>     active = true, -- optional
>     on =
>     {
>         -- at least one of these:
>         customEvents = { ... },
>         devices = { ... },
>         groups = { ... },
>         httpResponses = { ... },
>         scenes = { ... },
>         security = { ... },
>         shellCommandResponses = { ... },
>         system = { ... },
>         timer = { ... },
>         variables = { ... },
> 
>     },
>     data = { ... }, -- optional
>     logging = { ... }, -- optional
>     execute = function(domoticz, item, triggerInfo)
>     -- your code here
>     end
> }
> ```
> 
> But would be great if a Devices sections were added where you could specifiy your devices which are automatically created (and also automatically deleted if the script is deleted or the definition is removed), so something like:
> 
> ```
> return
> {
>     active = true, -- optional
>     VirtualDevices = { ... } 
>     on =
>     {
>         -- at least one of these:
>         customEvents = { ... },
>         devices = { ... },
>         groups = { ... },
>         httpResponses = { ... },
>         scenes = { ... },
>         security = { ... },
>         shellCommandResponses = { ... },
>         system = { ... },
>         timer = { ... },
>         variables = { ... },
> 
>     },
>     data = { ... }, -- optional
>     logging = { ... }, -- optional
>     execute = function(domoticz, item, triggerInfo)
>     -- your code here
>     end
> }
> ```
> 
> Where in the customdevices sections, you could specify the kind of virtual devices you would be using in the script.

This would be a nice feature too! But it should be possible, to still have manual functions to create the sensors, because - at least in my cases - the type of sensor to be created depends on some information which is gathered during the script run.

**Comment by akamming** [comment]
> This would be a nice feature too! But it should be possible, to still have manual functions to create the sensors, because - at least in my cases - the type of sensor to be created depends on some information which is gathered during the script run.

 a check, i understand now. You could use the info of the 1st 3 paragraphs here to do it :  [json API](https://wiki.domoticz.com/Domoticz_API/JSON_URL%27s#Create,_delete,_rename,_action_and_disable_commands), although i agree it would be a much nicer solution to have functions in dzVents which do this for you...

**Comment by elxsi** [comment]
Hi akamming,

yes, this is how I've implemented it, but it's not really a nice or clean solution.

Thanks!

**Comment by akamming** [comment]
> @akamming as a python plugin author, why do you not use a python plugin for creating and maintaining devices?

That does not fulfil all use cases: 

E.g. one of my dzvents scripts manages my energy consumption:  The logic is basically this:
- Trigger: P1 meter
- Action: 
  - retrieve current value from solar inverter using modbus (so i have them at the same moment as the P1)
  - calculate actual energy consumption by substracting the solar inverter values (energy delivered) from the P1 values 
  - update the virtual sensor for actual energy consumption

Or another one is that i calculate how much my car charges at home:
- trigger: battery percentage of my car
- evaluate conditions: Is the car at home? Is the percentage going up?
- If so: add a calculated number of kwh's to my charged at home virtual energy counter

I cannot build these in the python framework, cause python framework cannot be triggered by sensors outside the python script.  Furthermore: These kind of sensors/scripts are very easy to build in dzvents, it would be a lot of overhead to use the python framework for this.

**Comment by waltervl** [comment]
My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents. 

I hope you will find a developer that is willing to program and maintain this as it when I look how much effort the python framework took this is for sure not an easy task.

**Comment by akamming** [comment]
> My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents.
> 
> I hope you will find a developer that is willing to program and maintain this as it when I look how much effort the python framework took this is for sure not an easy task.

i think you are underestimating the power of adding such tools to domoticz.  I recently changed my setup so that is a combi of Domoticz and HAss and found out Hass has something call "template sensors", which are basically virtual sensors, but you can add code to it (called "template"). This is extremely powerful and unlocks a lot of use cases. 

And yes it's true you can do the same with adding virtual sensors in domoticz and than adding scripts to maintain them, it's just that this mechanism is so much more powerful and would be a good addition.

Nevertheless I do agree someone needs to maintain...  Hopefully some devs step up to the task... cause it would be a very good inprovement to the product...

**Comment by elxsi** [comment]
> My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents.

The thing is, for example in my case, that I often have LuA scripts which are really simple, query some devices for data and add the values to domoticz virtual sensors, but this devices usually return multiple values for different kind of attached sensors, so this means I have to manually create dozens of sensors. Due to the simplicity of the LuA code, it is often not worth the effort of developing a complete python plugin.

> I hope you will find a developer that is willing to program and maintain this as it when I look how much effort the python framework took this is for sure not an easy task.

Yes, of course, the development effort is for sure not an easy one.

**Comment by waltervl** [comment]
> > My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents.
> > I hope you will find a developer that is willing to program and maintain this as it when I look how much effort the python framework took this is for sure not an easy task.
> 
> i think you are underestimating the power of adding such tools to domoticz. I recently changed my setup so that is a combi of Domoticz and HAss and found out Hass has something call "template sensors", which are basically virtual sensors, but you can add code to it (called "template"). This is extremely powerful and unlocks a lot of use cases.
> 
> And yes it's true you can do the same with adding virtual sensors in domoticz and than adding scripts to maintain them, it's just that this mechanism is so much more powerful and would be a good addition.
> 
> Nevertheless I do agree someone needs to maintain... Hopefully some devs step up to the task... cause it would be a very good inprovement to the product...

The concept of template sensors is something completely different then hwat is suggested by the topic starter. And also not needed for Domoticz as we have dzvents... Create a virtual sensor manually and add dzvents scripts connecting to it.

**Comment by waltervl** [comment]
> > My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents.
> 
> The thing is, for example in my case, that I often have LuA scripts which are really simple, query some devices for data and add the values to domoticz virtual sensors, but this devices usually return multiple values for different kind of attached sensors, so this means I have to manually create dozens of sensors. Due to the simplicity of the LuA code, it is often not worth the effort of developing a complete python plugin.
> 
Why would doing some scripting work create dozens of devices? I don't understand...

**Comment by elxsi** [comment]
> > > My 2 cents as a user, just create manually some virtual devices occasionally does not need a full device creation functionality in dVents.
> > 
> > 
> > The thing is, for example in my case, that I often have LuA scripts which are really simple, query some devices for data and add the values to domoticz virtual sensors, but this devices usually return multiple values for different kind of attached sensors, so this means I have to manually create dozens of sensors. Due to the simplicity of the LuA code, it is often not worth the effort of developing a complete python plugin.
> 
> Why would doing some scripting work create dozens of devices? I don't understand...

For example, I'm retrieving data from a Heatpump via a special protocol based on UDP, the heatpump has different sensors like temperatures, current, wattage, internal status registers like current state / diagnostics, setpoints etc (about 30 sensors in total). The script is executed once every minute and the sensors are read and the virtual devices are updated. Some data is also based on calculations during the script run like the COP value. Of course, this could also be achieved with a python plugin, but the overhead is much higher.

**Comment by waltervl** [comment]
Yes, this is typical a python plugin use case. It could even be that your heatpump is already supported by a Python plugin

**Comment by elxsi** [comment]
This was just one example. The thing is, that development of a python plugin is a lot more effort and still has some disadvantages. LuA is simple to use, dzVents provides a really nice interface and simple to use functions, but it lacks the creation of virtual sensors, so I still think this would be a great feature.

**Comment by waltervl** [comment]
There are lots of python plugins available to learn. Also adding Device Management to Dzvents will make dzvents scripts more complex (and similar to python plugins).

**Comment by elxsi** [comment]
It's not about learning, I've already developed some python plugins. But it's still more time consuming than just writing some simple LuA stuff. I also don't see the point, why adding a new function for device creation would make dzVents scripts more complex.

Nevertheless, I see that there's not much interest from your side regarding this. So feel free to close the issue, I can live without this (still thinking it would be a nice addition though) and maybe someone will look into this at a later time.

Thanks!

**Comment by waltervl** [comment]
It is not only creating a device, you also have to check if the device is already there. What happens if you changed the name of the device? What happens if you changed the characteristics of the device? What happens if you disabled the device? 

So you need device management, not only a simple device creation function.

If you only want to create a device it is very easy for you to create a dzvents helper function for this.

**Comment by elxsi** [comment]
Yes, the logic of the dzVents internals must of course handle this, so maybe I misunderstood you, but I've taken "will make dzvents scripts more complex" from the user perspective.

**Comment by waltervl** [comment]
Dzvents cannot  do device management without the user deciding what to do with devices changes. So if a user decides to create devices by dzvents it also has to define something for the other device management actions. This cannot be handled by dzvents internals only. So from a user perspective the scripts will become more complex if the user decides to use device management.

**Comment by elxsi** [comment]
Well, I still don't think that this makes it really difficult from a user perspective.

local dev = dz.createDevice(...)

If device creation fails (if device already exists or because of other errors), return nil (and/or throw/print an error), otherwise return the created device. Same with updating device parameters:

local dev = dz.updateDevice(currentDev, ...)

Return new device details or nil.

So all the user has to do is to check if dev is nil, shouldn't be a major issue I think.

**Comment by gizmocuz** [comment]
I'm not against this, could you add these functions and also extend the dzVents tests for these?

**Comment by gizmocuz** [comment]
Otherwise, just add a dummy hardware and create your sensors
